### PR TITLE
Improve plot recipe for spatial data

### PR DESCRIPTION
**Pull request by LakshyaKhatri**
fixes https://github.com/JuliaEarth/GeoStats.jl/issues/74

### CODE REVIEW

File: src/plotrecipes/data.jl [hunk 1]
```diff
@@ -2,7 +2,7 @@
 # Licensed under the ISC License. See LICENSE in the project root.
 # ------------------------------------------------------------------
 
-@recipe function f(sdata::AbstractData; variables=nothing)
+@recipe function f(sdata::AbstractData, variables=nothing)
```
Comment (juliohm) [inline src/plotrecipes/data.jl:5]: Nice, I would even rename the argument to `vars` to avoid conflicting with the `GeoStatsBase.variables` function. :)

Comment (LakshyaKhatri) [reply src/plotrecipes/data.jl:5]: Oh, you made that change recently, I almost missed it. Thanks :smile:

File: src/plotrecipes/data.jl [hunk 2]
```diff
@@ -2,21 +2,21 @@
 # Licensed under the ISC License. See LICENSE in the project root.
 # ------------------------------------------------------------------
 
-@recipe function f(sdata::AbstractData; variables=nothing)
+@recipe function f(sdata::AbstractData, vars=nothing)
   # retrieve underlying domain
   sdomain = domain(sdata)
 
   # valid variables
   validvars = name.(GeoStatsBase.variables(sdata))
```
Comment (juliohm) [inline src/plotrecipes/data.jl:10]: given the new name I think now we can remove the `GeoStatsBase` prefix 👌

**Comment by LakshyaKhatri** [comment]
@juliohm, I don't understand how `isvalid()` is being used in the current context, would it be possible for you to provide some pointers?

**Comment by juliohm** [comment]
Of course @LakshyaKhatri , thanks for asking. The `isvalid` function will be deprecated (see https://github.com/JuliaEarth/GeoStats.jl/issues/42). It is been used in the plot recipe to check if a value is missing and then replace the value by `NaN`. We should get rid of isvalid here is possible.

Consider the following spatial data:

```julia
using GeoStatsBase
using CategoricalArrays

# image layers
A = rand(3,3)
B = rand(Int, 3, 3)
C = string.(B)
D = categorical(reshape(1:9, 3, 3))

# images with missings
E = reshape([1,2,3,4,5,6,7,8,missing], 3, 3)
F = reshape(vcat(string.(1:8), missing), 3, 3)
G = categorical(F)

# spatial data
sdata = georef((;A, B, C, D, E, F, G))
```

For columns A through D, we have no missing values, but we have categorical arrays. In the case of categorical arrays, we need to extract the codes and plot as integers with colors: `levelcode.(D)`. For columns E through G, we have missing values so there is an additional preprocessing layer to replace the `missing` by `NaN` for plotting given that the plotting ecosystem handles `NaN` already. The good news is that the latest categorical arrays `levelcode.(G)` for example will preserve the missing values.

I would implement this "column preprocessing" or "extraction" in a separate auxiliar function like `plotspec(column)`. This `plotspec` function would take care of replacing `missing` by `NaN` and extracting the level codes for plotting. It would also return the elements of the legend in case of categorical arrays (e.g. the corresponding strings).

Does it make sense?

**Comment by juliohm** [comment]
The general message here is that the Plots.jl system doesn't take care of the scientific type of the array, and so we have to manually convert things before plotting. Ideally, the plotting system and recipes would be described in terms of ScientificTypes.jl so any array with known scientific type could be plotted directly (categorical, continuous, counts, etc.).

**Comment by LakshyaKhatri** [comment]
Yes, there are less clouds now :D

**Comment by LakshyaKhatri** [comment]
I forgot to handle categorical arrays, let me fix that.

**Comment by juliohm** [comment]
I would first use `levelcode(v)` to get the array with integers and missing and then postprocess this array to replace missing by NaN. We could even eliminate this counter intuitive `map` syntax.

**Comment by LakshyaKhatri** [comment]
So, the flow will be something like:

- Itterate over each column in `sdata`
- Check if the column is a `CategoricalArray`
- if yes then do `levelcode.(column)`
- replace `missing` with `NaN` in that column

**Comment by juliohm** [comment]
Perfect

**Comment by LakshyaKhatri** [comment]
@juliohm, this is adding a dependency for `levelcode()`. Is it okay?

**Comment by juliohm** [comment]
Oh `levelcode` isn't exported by CategoricalArrays.jl already?

**Comment by LakshyaKhatri** [comment]
Yes, it is exported by CategoricalArrays.jl. I was just asking that is it okay to do a `using CategoricalArrays`  in our recipe file?

**Comment by juliohm** [comment]
One thing we could consider in the future is treating this issue of categorical, continuous, count, etc in the Plots.jl ecosystem directly. That is, instead of doing this manipulation by hand here, we could open PRs in RecipesBase.jl to do this preprocessing in the plots based on the scientific type of the data. For example, calling `heatmap(rand(10,10))` should produce a plot with continuous variables and `heatmap(categorical(rand(Int,10,10))` should produce a plot with categorical variables with a legend for the levels. This is a major change though that requires a deep understanding of the RecipesBase.jl system I guess.

**Comment by juliohm** [comment]
You can add the `levelcode` entry in `src/GeoStatsBase.jl`, it is already a dependency

**Comment by LakshyaKhatri** [comment]
I need to update the tests now.

**Comment by juliohm** [comment]
Yes, that would be awesome! Please let me know if you need some help, there is a decent number of visual tests.

File: src/plotrecipes/data.jl [hunk 3]
```diff
@@ -2,30 +2,26 @@
...
-      if isvalid(sdata, ind, var)
-        v = sdata[ind,var]
-        v isa Number ? v : get(v)
-      else
-        NaN
-      end
-    end
+    vals = sdata[var]
+    vals = vals isa CategoricalArray ? levelcode.(vals) : vals
+    coalesce.(vals, NaN)  # replace missing values with NaN
```
Comment (juliohm) [inline src/plotrecipes/data.jl:23]: perhaps is it more efficient and explicit to do

```julia
vals = replace(vals, missing => NaN)
```

**Comment by LakshyaKhatri** [comment]
Hello @juliohm, sorry for the late response.
Do I have to address the tests that are failing right now?

**Comment by juliohm** [comment]
Thank you @LakshyaKhatri :heart: no need to fix the test, it is unrelated to the changes proposed here :)

I appreciate your time and professional PRs. Squashing and merging :100:

**Pull request merged**